### PR TITLE
add: ci tooling for security, type, lint checks

### PR DESCRIPTION
**Pull request by ayhamthemayhem**
Extend existing GitHub actions workflow with the recommendations made by this [security audit](https://github.com/trailofbits/publications/blob/master/reviews/2023-09-pypi-warehouse-securityreview.pdf)
- ~~black for auto-formatting~~ ( Was already in there)
- [x] mypy for type-checking
- [x] bandit for automatic detection of common security issues
- ~~Include test runs~~ (There are no meaningful tests in the repo at the moment)

### CODE REVIEW

**Comment by ewdurbin** [comment]
This conflicts with #81, which removes tox as a dependency locally and simplifies dev to just requiring docker/docker-compose.

Would you be opposed to seing #81 merged and rebasing on that?

**Comment by ayhamthemayhem** [comment]
All good I can adjust after #81 gets merged :)

**Comment by ewdurbin** [comment]
👍🏼, I've merged #81

One note, is to add the new deps to requirements/dev.in rather than requirements/base.in. I noticed that this PR also upgraded _all_ dependencies which is probably not ideal in one-shot.

**Comment by ayhamthemayhem** [comment]
Good to know ! I noticed that the deps where updated also this is the first time for me working on a python project and there still so much to learn here :).

Hopefully I can adjust land this PR by early next week.

**Comment by ewdurbin** [comment]
Cool, only other note is that both mypy and bandit support configuration via pyproject.toml:

- https://mypy.readthedocs.io/en/stable/config_file.html#example-pyproject-toml
- https://bandit.readthedocs.io/en/latest/config.html

I think we should prefer that over having a bunch of various config files added.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -1,23 +1,39 @@
...
+  pull_request:
+    branches:
+      - main
+
 jobs:
   test:
+    runs-on: ubuntu-latest
     strategy:
       matrix:
+        python-version: [3.12]
```
Comment (ewdurbin) [inline .github/workflows/ci.yml:16]: I'm not sure we want to run CI on a version that we're not running the app on:

https://github.com/cabotage/cabotage-app/blob/6b9b216f045bcaf742b14eca6da11237a19b664a/Dockerfile#L1

I think 3.11 is best for now until we target 3.12 for runtime.

File: Makefile [hunk 2]
```diff
@@ -1,8 +1,8 @@
 .PHONY: default \
-	lint reformat \
+	lint reformat type-check security-check \
 	start stop rebuild destroy \
 	migrate create-admin \
-	routes
+	routes install-deps install-tox
```
Comment (ewdurbin) [inline Makefile:5]: I think this is extraneous now.

**Comment by ewdurbin** [comment]
Hi @ayhamthemayhem! Can you address the comments above? I don't think we want to upgrade to Python 3.12 as part of this, and it appears that there is some vestigial references to tox after #81 was merged.

**Comment by ayhamthemayhem** [comment]
Python update was part of me trying to figure why there is a miss match in the dep hashes when I run the CI in our fork and that is caused by some deps being imported in `dev.txt` and `base.txt` by different packages but I will address it in my fork and I will bring it back to 3.11.
If you have some insight on how to resolve this issue please let me know :)

File: cabotage/server/models/projects.py [hunk 3]
```diff
@@ -84,7 +84,7 @@
 
 
 class Project(db.Model, Timestamp):
-    __versioned__ = {}
+    _versioned__: dict = {}
```
Comment (ewdurbin) [inline cabotage/server/models/projects.py:87]: all of the matching changes in this file remove the leading `_` which breaks sqlalchemy-continuum:

```
cabotage-app-1         | Traceback (most recent call last):
cabotage-app-1         |   File "/opt/cabotage-app/lib/python3.11/site-packages/gunicorn/workers/gthread.py", line 282, in handle
cabotage-app-1         |     keepalive = self.handle_request(req, conn)
cabotage-app-1         |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
cabotage-app-1         |   File "/opt/cabotage-app/lib/python3.11/site-packages/gunicorn/workers/gthread.py", line 334, in handle_request
cabotage-app-1         |     respiter = self.wsgi(environ, resp.start_response)
cabotage-app-1         |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
cabotage-app-1         |   File "/opt/cabotage-app/lib/python3.11/site-packages/sentry_sdk/integrations/flask.py", line 84, in sentry_patched_wsgi_app
cabotage-app-1         |     return SentryWsgiMiddleware(lambda *a, **kw: old_app(self, *a, **kw))(
cabotage-app-1         |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
cabotage-app-1         |   File "/opt/cabotage-app/lib/python3.11/site-packages/sentry_sdk/integrations/wsgi.py", line 109, in __call__
cabotage-app-1         |     reraise(*_capture_exception())
cabotage-app-1         |   File "/opt/cabotage-app/lib/python3.11/site-packages/sentry_sdk/utils.py", line 1640, in reraise
cabotage-app-1         |     raise value
cabotage-app-1         |   File "/opt/cabotage-app/lib/python3.11/site-packages/sentry_sdk/integrations/wsgi.py", line 102, in __call__
cabotage-app-1         |     response = self.app(
cabotage-app-1         |                ^^^^^^^^^
cabotage-app-1         |   File "/opt/cabotage-app/lib/python3.11/site-packages/sentry_sdk/integrations/flask.py", line 84, in <lambda>
cabotage-app-1         |     return SentryWsgiMiddleware(lambda *a, **kw: old_app(self, *a, **kw))(
cabotage-app-1         |                                                  ^^^^^^^^^^^^^^^^^^^^^^^
cabotage-app-1         |   File "/opt/cabotage-app/lib/python3.11/site-packages/flask/app.py", line 2213, in __call__
cabotage-app-1         |     return self.wsgi_app(environ, start_response)
cabotage-app-1         |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
cabotage-app-1         |   File "/opt/cabotage-app/lib/python3.11/site-packages/flask/app.py", line 2193, in wsgi_app
cabotage-app-1         |     response = self.handle_exception(e)
cabotage-app-1         |                ^^^^^^^^^^^^^^^^^^^^^^^^
cabotage-app-1         |   File "/opt/cabotage-app/lib/python3.11/site-packages/flask/app.py", line 2190, in wsgi_app
cabotage-app-1         |     response = self.full_dispatch_request()
cabotage-app-1         |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
cabotage-app-1         |   File "/opt/cabotage-app/lib/python3.11/site-packages/flask/app.py", line 1486, in full_dispatch_request
cabotage-app-1         |     rv = self.handle_user_exception(e)
cabotage-app-1         |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
cabotage-app-1         |   File "/opt/cabotage-app/lib/python3.11/site-packages/flask/app.py", line 1484, in full_dispatch_request
cabotage-app-1         |     rv = self.dispatch_request()
cabotage-app-1         |          ^^^^^^^^^^^^^^^^^^^^^^^
cabotage-app-1         |   File "/opt/cabotage-app/lib/python3.11/site-packages/flask_debugtoolbar/__init__.py", line 157, in dispatch_request
cabotage-app-1         |     return view_func(**req.view_args)
cabotage-app-1         |            ^^^^^^^^^^^^^^^^^^^^^^^^^^
cabotage-app-1         |   File "/opt/cabotage-app/lib/python3.11/site-packages/flask_login/utils.py", line 290, in decorated_view
cabotage-app-1         |     return current_app.ensure_sync(func)(*args, **kwargs)
cabotage-app-1         |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
cabotage-app-1         |   File "/opt/cabotage-app/src/cabotage/server/user/views.py", line 288, in project_application
cabotage-app-1         |     view_releases=version_class(Release)
cabotage-app-1         |                   ^^^^^^^^^^^^^^^^^^^^^^
cabotage-app-1         |   File "/opt/cabotage-app/lib/python3.11/site-packages/sqlalchemy_continuum/utils.py", line 123, in version_class
cabotage-app-1         |     manager = get_versioning_manager(model)
cabotage-app-1         |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
cabotage-app-1         |   File "/opt/cabotage-app/lib/python3.11/site-packages/sqlalchemy_continuum/utils.py", line 30, in get_versioning_manager
cabotage-app-1         |     raise ClassNotVersioned(cls.__name__)
cabotage-app-1         | sqlalchemy_continuum.exc.ClassNotVersioned: Release
```

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -1,23 +1,39 @@
...
     strategy:
       matrix:
+        python-version: [3.11]
         include:
           - name: Lint
-            command: bin/lint
-    runs-on: ubuntu-latest
+            command: make lint
+          - name: Reformat
+            command: make reformat
```
Comment (ewdurbin) [inline .github/workflows/ci.yml:21]: I think we should remove this call since `make reformat` is most often only useful locally. Since black will happily reformat the code in place in CI, this isn't actually checking that code is reformatted.

Perhaps a step in the `bin/lint` to run `black --check` would be best.

**Review by ewdurbin — CHANGES_REQUESTED** [review]
Regarding dependencies, I think maintaining a separation between dev dependencies is important. In order to handle the scenario where dev resolves an incompatible dependency, consider setting a constraint on that dependency in dev.in to match what base.txt has resolved.

File: pyproject.toml [hunk 5]
```diff
@@ -0,0 +1,6 @@
+[tool.bandit]
+exclude_dirs = ["bin", "lib"]
+skips =["B105","B106","B603","B404"]
```
Comment (ewdurbin) [inline pyproject.toml:3]: Rather than blanket skips, consider using `# nosec` comments, per bandit docs: https://bandit.readthedocs.io/en/latest/config.html#suppressing-individual-lines

File: cabotage/server/models/auth.py [hunk 6]
```diff
@@ -52,7 +52,7 @@ def __hash__(self):
 
 
 class User(db.Model, FsUserMixin):
-    __versioned__ = {
+    ___versioned__: dict = {
```
Comment (ewdurbin) [inline cabotage/server/models/auth.py:55]: these are now `___versioned__` and should be `__versioned__`.

File: cabotage/server/models/resources.py [hunk 7]
```diff
@@ -6,7 +6,7 @@
 
 
 class Resource(db.Model, Timestamp):
-    __versioned__ = {}
+    ___versioned__: dict = {}
```
Comment (ewdurbin) [inline cabotage/server/models/resources.py:9]: same concern with `___versioned__` instead of `__versioned__`

File: cabotage/server/user/views.py [hunk 8]
```diff
@@ -845,8 +845,8 @@ def project_application_configuration_delete(
                 app_slug,
                 configuration,
             )
-        except Exception:
-            pass  # No, we should def not do this
+        except Exception :
+            raise
```
Comment (ewdurbin) [inline cabotage/server/user/views.py:849]:
```suggestion
            raise  # TODO: Coordinate configuration deletion between db and config_writer
```

File: requirements/base.in [hunk 9]
```diff
@@ -17,22 +18,25 @@ flask-migrate
 flask-nav3
 flask-security-too
 flask-sock
+urllib3==2.2.2
 flask-sqlalchemy
+greenlet
```
Comment (ewdurbin) [inline requirements/base.in:23]: why was this added as a direct dependency? we don't use it directly and can allow it to be a transitive dependency from sqlalchemy.

**Review by ewdurbin — CHANGES_REQUESTED** [review]

File: requirements/base.in (hunk 9 above)
Comment (ayhamthemayhem) [reply requirements/base.in:23]: I was getting this error and adding it fixed it for me 
```
6.990 Collecting greenlet!=0.4.17 (from sqlalchemy==1.4.48->-r /tmp/requirements/base.txt (line 690))
6.990 ERROR: In --require-hashes mode, all requirements must have their versions pinned with ==. These do not:
6.990     greenlet!=0.4.17 from https://files.pythonhosted.org/packages/21/b4/90e06e07c78513ab03855768200bdb35c8e764e805b3f14fb488e56f82dc/greenlet-3.0.3-cp311-cp311-manylinux_2_17_aarch64.manylinux2014_aarch64.whl (from sqlalchemy==1.4.48->-r /tmp/requirements/base.txt (line 690))
```

File: bin/lint [hunk 10]
```diff
@@ -2,3 +2,4 @@
 
 black --check --diff .
 ruff check .
+black --check
```
Comment (ewdurbin) [inline bin/lint:5]: huh, it looks like line 3 should already be running black --check. I'm guessing what's occurred is that the CI run was reformatting before lint. this line can now be removed.

File: requirements/base.in (hunk 9 above)
Comment (ayhamthemayhem) [reply requirements/base.in:23]: When I remove greenlet from base.in and recompile the deps its breaking for me 
` pip-compile --generate-hashes --output-file=requirements/base.txt requirements/base.in`

Comment (ayhamthemayhem) [reply requirements/base.in]: are you able to see the error in this build ?
https://github.com/neighbourhoodie/cabotage-app/actions/runs/10094419278/job/27912224625?pr=2

Comment (ewdurbin) [reply requirements/base.in]: it looks like that is because you manually removed greenlet from requirements/base.txt as well.

when I remove `greenlet` from requirements/base.in, then run `make requirements/base.txt` the result is fine.

Comment (ayhamthemayhem) [reply requirements/base.in]: I am not manually removing it from base.txt but when I use pip-compile it gets removed from base.txt automatically and I thought this is the required behavior here

Comment (ewdurbin) [reply requirements/base.in]: it looks like you may be running pip-compile locally instead of using the make target. it is unexpected that the transitive dependency would be removed, since sqlalchemy does have a dependency on greenlet for some conditions: https://github.com/sqlalchemy/sqlalchemy/blob/rel_1_4_48/setup.cfg#L45

if you're running pip-compile in an environment that is excluded, it will be dropped.

using the make target runs pip-compile in an environment that will replicate our CI and deployment targets.

**Review by ewdurbin — APPROVED** [review]

**Comment by ewdurbin** [comment]
Thanks @ayhamthemayhem! One note for future PRs. We've enabled "Squash and merge" for pull requests so that there's no need to continually re-base and force-push to have things be a single commit. The entire changeset from the PR will be merged to `main` as a single commit.

Force pushing makes review more difficult than it needs to be as changes are pushed, leaving comments made during review inscrutable.
